### ISSUE TITLE
Self referential foreign key causes subsequent alters to fail

### ISSUE THREAD

**Issue by damienhollis** [cockroachdb/cockroach#14586]
**BUG REPORT**

Build Tag:    beta-20170330
Build Time:   2017/03/30 15:14:06
Distribution: CCL
Platform:     linux amd64
Go Version:   go1.8
C Compiler:   gcc 6.3.0
Build SHA-1:  f0f5783cc67f9cda16cedf1b88ca0691b0b65884
Build Type:   release

**Issue**

I was trying to create tables/foreign keys as follows:
`CREATE TABLE A (ID BIGSERIAL NOT NULL, SELF_ID INT, B_ID INT NOT NULL, PRIMARY KEY (ID));`
`CREATE TABLE B (ID BIGSERIAL NOT NULL, PRIMARY KEY (ID));`
`ALTER TABLE A ADD CONSTRAINT FK_SELF_ID FOREIGN KEY (SELF_ID) REFERENCES A;`

Up to this point everything works.  But the following causes an error:

`ALTER TABLE A ADD CONSTRAINT FK_B FOREIGN KEY (B_ID) REFERENCES B;`
"pq: missing fk back reference to A.A_auto_index_FK_SELF_ID from A.primary"

Also, at this point anything I've tried related to changing table A fails:
`DROP TABLE A;`
`ALTER TABLE A DROP CONSTRAINT FK_SELF_ID;`
all result in "pq: missing fk back reference to A.A_auto_index_FK_SELF_ID from A.primary"

I'm happy to take a look at the code and see if I can implement a patch, if somebody wants to point me in the right direction.

**Comment by eldos-dl** [comment]
I am still seeing this issue
```
CREATE TABLE a (id BIGSERIAL NOT NULL, self_id INT, b_id INT UNIQUE NOT NULL, PRIMARY KEY (id));
ALTER TABLE a ADD CONSTRAINT fk_self_id FOREIGN KEY (self_id) REFERENCES a(b_id);
DROP TABLE a;
```
This throws `pq: missing fk back reference to a.a_auto_index_fk_self_id from a.a_c_id_key` error.

**Comment by dt** [comment]
@eldos-dl Thanks for tracking that down! That error is actually hiding a much more serious one -- the FK requires an index on both the referencing and referenced columns, and since there is no index on `self_id`, it is (very incorrectly) trying to be helpful just making one. Adding the needed index on the fly like that is fine during CREATE TABLE, but is not okay here, where there may be existing data.

I just put together a fix in #16175 -- you'll still see an error if you try to run those, but it will be from the `ALTER` and can be avoided by running `CREATE INDEX ON a (self_id);` before the `ALTER` (or adding the index to the `CREATE TABLE` of course).
